structure_tests/drivers/dockerDriver.go

```diff
@@ -0,0 +1,298 @@
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drivers
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"context"
+
+	"bufio"
+
+	"github.com/GoogleCloudPlatform/runtimes-common/structure_tests/types/unversioned"
+	"github.com/GoogleCloudPlatform/runtimes-common/structure_tests/utils"
+	"github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	OriginalImage string
+	CurrentImage  string
+	cli           docker.Client
+}
+
+func (d DockerDriver) Info() string {
+	return fmt.Sprintf("DockerDriver:\nOriginalImage: %s\nCurrentImage: %s\ncli: %s\n",
+		d.OriginalImage, d.CurrentImage, d.cli)
+}
+
+func NewDockerDriver(image string) DockerDriver {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		panic(err)
+	}
+	return DockerDriver{
+		OriginalImage: image,
+		CurrentImage:  image,
+		cli:           *newCli,
+	}
+}
+
+func (d DockerDriver) ProcessCommand(t *testing.T, envVars []unversioned.EnvVar, fullCommand []string,
+	shellMode bool, checkOutput bool) (string, string, int) {
+
+	if len(fullCommand) == 0 {
+		t.Logf("empty command provided: skipping...")
+		return "", "", -1
+	}
+	env := []string{}
+	if len(envVars) > 0 {
+		for _, envVar := range envVars {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	//reset image for next test
+	d.CurrentImage = d.OriginalImage
+	return stdout, stderr, exitCode
+}
+
+func (d DockerDriver) SetEnvVars(t *testing.T, vars []unversioned.EnvVar) []unversioned.EnvVar {
+	// currently, this does not support substitution, so setting env vars is broken
+	// need to find a way to retrieve the existing values from the container before setting
+	if len(vars) == 0 {
+		return nil
+	}
+	env := []string{}
+	for _, envVar := range vars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	newImage := d.runAndCommit(t, env, nil)
+	// since these are global envvars, just overwrite the original image
+	d.OriginalImage = newImage
+	d.CurrentImage = d.OriginalImage
+	return nil
+}
+
+func (d DockerDriver) ResetEnvVars(t *testing.T, vars []unversioned.EnvVar) {
+	// since the container will be destroyed after the tests, this is a noop
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d DockerDriver) retrieveFile(t *testing.T, path string, directory bool) (string, error) {
```

Comment:
Can we extract a file from an image, or only from a container?

Reply:
I pulled it from a running container because I don't think it's possible with the client, but I have plans to write another driver that doesn't use docker and will probably do tar unpacking, in which case we'll be able to get files from just images.